src/parquet/column/levels-test.cc

```diff
@@ -0,0 +1,122 @@
+
+  void TearDown() {}
+
+ protected:
+  std::vector<int16_t> input_levels_;
+  std::vector<int16_t> output_levels_;
+  std::vector<uint8_t> bytes_;
+};
+
+// test levels with maximum bit-width from 1 to 8
+// increase the repetition count for each iteration by a factor of 2
+TEST_F(TestLevels, TestEncodeDecodeLevels) {
+  LevelEncoder encoder;
+  LevelDecoder decoder;
+  int min_repeat_factor = 0;
+  int max_repeat_factor = 7; // 128
+  int max_bit_width = 8;
+  parquet::Encoding::type encodings[2] = {parquet::Encoding::RLE,
+      parquet::Encoding::BIT_PACKED};
+
+  // for each encoding
+  for (int encode = 0; encode < 2; encode++) {
+    int levels_count = 0;
+    int num_levels = 0;
+    parquet::Encoding::type encoding = encodings[encode];
+    //BIT_PACKED requires a sequence of atleast 8
+    if ((encoding == parquet::Encoding::BIT_PACKED)) min_repeat_factor = 3;
+
+    // for each maximum bit-width
+    for (int bit_width = 1; bit_width <= max_bit_width; bit_width++) {
+      int num_levels_per_width = ((2 << max_repeat_factor) - (1 << min_repeat_factor));
+      num_levels = (bit_width + 1) * num_levels_per_width;
+      Init(num_levels);
+      // find the maximum level for the current bit_width
+      int max_level = (1 << bit_width) - 1;
+      int total_count = 0;
+      // for each repetition count upto max_repeat_factor
+      for (int repeat = min_repeat_factor; repeat <= max_repeat_factor; repeat++) {
+        // repeat count increase by a factor of 2 for every iteration
+        int repeat_count = (1 << repeat);
+        // generate levels for repetition count upto the maximum level
+        int value = 0;
+        int bwidth = 0;
+        while (value <= max_level) {
+          for (int i = 0; i < repeat_count; i++) {
+            input_levels_[total_count++] = value;
+          }
+          value = (2 << bwidth) - 1;
+          bwidth++;
+        }
+      }
+      // levels are intialized
+      ASSERT_EQ(num_levels, total_count);
+      // start encoding and decoding
+      if (encoding == parquet::Encoding::RLE) {
+        // leave space to write the rle length value
+        encoder.Init(encoding, max_level, num_levels,
+            bytes_.data() + sizeof(uint32_t), sizeof(uint8_t) * 2 * num_levels);
+
+        levels_count = encoder.Encode(num_levels, input_levels_.data());
+        (reinterpret_cast<uint32_t*>(bytes_.data()))[0] = encoder.len();
+
+      } else {
+        encoder.Init(encoding, max_level, num_levels,
+            bytes_.data(), sizeof(uint8_t) * 2 * num_levels);
+        levels_count = encoder.Encode(num_levels, input_levels_.data());
+      }
+
+      ASSERT_EQ(levels_count, num_levels);
+
+      decoder.Init(encoding, max_level, num_levels, bytes_.data());
+      levels_count = decoder.Decode(num_levels, output_levels_.data());
+
+      ASSERT_EQ(levels_count, num_levels);
+
+      for (int i = 0; i < num_levels; i++) {
+        EXPECT_EQ(input_levels_[i], output_levels_[i]);
+      }
```

Comment:
can you turn this into a reusable testing function that accepts const std::vector<int16_t>& levels and the encoding, then verifies that the values survive the round trip? This way we can more easily expand the test cases and different types of data generation in the future.